### PR TITLE
Python 3.10

### PR DESCRIPTION
**Pull request by graingert**
combines https://github.com/dask/distributed/pull/5778 and https://github.com/dask/distributed/pull/5353

- [x] Closes https://github.com/dask/distributed/issues/5460
- [x] Tests added / passed
- [x] Passes `pre-commit run --all-files`

### CODE REVIEW

File: .github/workflows/tests.yaml [hunk 1]
```diff
@@ -65,12 +65,6 @@ jobs:
         shell: bash -l {0}
         run: conda config --show
 
-      - name: Install stacktrace
-        shell: bash -l {0}
-        # stacktrace for Python 3.8 has not been released at the moment of writing
-        if: ${{ matrix.os == 'ubuntu-latest' && matrix.python-version < '3.8' }}
-        run: mamba install -c conda-forge -c defaults -c numba libunwind stacktrace
```
Comment (jakirkham) [inline .github/workflows/tests.yaml:72]: xref: https://github.com/dask/distributed/issues/5829

File: distributed/tests/test_client.py [hunk 2]
```diff
@@ -6461,24 +6461,15 @@ async def f(stacklevel, mode=None):
     assert "cdn.bokeh.org" in data
 
 
-@gen_cluster(nthreads=[])
-async def test_client_gather_semaphore_loop(s):
-    async with Client(s.address, asynchronous=True) as c:
-        assert c._gather_semaphore._loop is c.loop.asyncio_loop
```
Comment (jakirkham) [inline distributed/tests/test_client.py:6467]: Is there a reason this test is dropped?

File: distributed/tests/test_client.py [hunk 3]
```diff
@@ -6461,24 +6461,15 @@ async def f(stacklevel, mode=None):
...
     seq = c.map(inc, range(5))
     ac = as_completed(seq)
+    async for _ in ac:
+        pass
     assert ac.condition._loop == c.loop.asyncio_loop
 
 
-def test_client_connectionpool_semaphore_loop(s, a, b):
-    with Client(s["address"]) as c:
-        assert c.rpc.semaphore._loop is c.loop.asyncio_loop
```
Comment (jakirkham) [inline distributed/tests/test_client.py:6479]: Is there a reason this test is dropped?

File: distributed/tests/test_client.py [hunk 4]
```diff
@@ -6461,24 +6461,15 @@ async def f(stacklevel, mode=None):
...
-    async with Client(s.address, asynchronous=True) as c:
-        assert c._gather_semaphore._loop is c.loop.asyncio_loop
-
-
 @gen_cluster(client=True)
 async def test_as_completed_condition_loop(c, s, a, b):
     seq = c.map(inc, range(5))
     ac = as_completed(seq)
+    async for _ in ac:
+        pass
```
Comment (jakirkham) [inline distributed/tests/test_client.py:6469]: What is the purpose of this change?

**Review by jakirkham — COMMENTED** [review]
Thanks for working on this Thomas! 😄

Had a few questions below 🙂

File: distributed/tests/test_client.py (hunk 4 above)
Comment (graingert) [reply distributed/tests/test_client.py:6469]: on python 3.10 async resources like asyncio.Condition bind the loop late: https://github.com/python/cpython/blob/ba76f901923d80ad9b24bb1636aa751d55e0c768/Lib/asyncio/mixins.py#L28

by iterating `ac` we're interacting with the `ac.condition` so it gets bound to a loop

Comment (jakirkham) [reply distributed/tests/test_client.py:6469]: Maybe worth including a comment then?

File: distributed/profile.py [hunk 5]
```diff
@@ -67,9 +67,24 @@ def repr_frame(frame):
     return text + "\n\t" + line
 
 
+class LazyFrameInfo:
+    def __init__(self, frame):
+        self._frame
```
Comment (graingert) [inline distributed/profile.py:72]:
```suggestion
        self._frame = frame
```

File: distributed/profile.py [hunk 6]
```diff
@@ -59,21 +60,40 @@ def identifier(frame):
...
+    f_lineno = _f_lineno(frame)
+    text = f'  File "{co.co_filename}", line {f_lineno}, in {co.co_name}'
+    line = linecache.getline(co.co_filename, f_lineno, frame.f_globals).lstrip()
     return text + "\n\t" + line
 
 
 def info_frame(frame):
     co = frame.f_code
-    line = linecache.getline(co.co_filename, frame.f_lineno, frame.f_globals).lstrip()
+    f_lineno = _f_lineno(frame)
```
Comment (graingert) [inline distributed/profile.py:91]: frames can have a `frame.f_lineno=None` In Python 3.10 this seems to be much more likely eg when dispatching a JUMP_ABSOLUTE: https://github.com/python/cpython/commit/f5e97b72fecff9b9ced35704ec5e6cad29e2825d#diff-ebc983d9f91e5bcf73500e377ac65e85863c4f77fd5b6b6caf4fcdf7c0f0b057R2777

we can calculate a nearby line with `frame.f_lasti` and `dis.findlinestarts`

Comment (graingert) [reply distributed/profile.py:91]: a smaller reproducer for this issue is here: https://gist.github.com/graingert/f5e81264dd3c3ab8a705809096fcc283#file-demo-py-L27


related to the logger gaining ~thousand DequeHandlers over the course of the test suite https://github.com/dask/distributed/issues/5973

File: distributed/tests/test_client.py [hunk 7]
```diff
@@ -6461,24 +6461,15 @@ async def f(stacklevel, mode=None):
...
-async def test_client_gather_semaphore_loop(s):
-    async with Client(s.address, asynchronous=True) as c:
-        assert c._gather_semaphore._loop is c.loop.asyncio_loop
-
-
 @gen_cluster(client=True)
 async def test_as_completed_condition_loop(c, s, a, b):
     seq = c.map(inc, range(5))
     ac = as_completed(seq)
+    async for _ in ac:
```
Comment (graingert) [inline distributed/tests/test_client.py:6468]:
```suggestion
    # consume the ac so that the ac.condition is bound to the loop on py3.10+
    async for _ in ac:
```

File: distributed/tests/test_client.py (hunk 3 above)
Comment (graingert) [reply distributed/tests/test_client.py:6479]: the on py3.10+ `c.rpc.semaphore` doesn't bind to the eventloop until the semaphore is exhausted and `c.rpc.semaphore.acquire()` blocks so I decided it was too difficult to get the test to pass and so not worth the maintenance to keep it around as skipped on py3.10+ until 3.9 EOL

File: distributed/tests/test_client.py (hunk 2 above)
Comment (graingert) [reply distributed/tests/test_client.py]: I've restored these as skipped on 3.10+

File: distributed/utils_test.py [hunk 8]
```diff
@@ -739,13 +739,16 @@ def cluster(
...
+                async def wait_for_workers():
+                    async with rpc(saddr, **rpc_kwargs) as s:
+                        while True:
+                            nthreads = await s.ncores()
+                            if len(nthreads) == nworkers:
+                                break
+                            if time() - start > 5:
+                                raise Exception("Timeout on cluster creation")
+
+                loop.run_sync(wait_for_workers)
```
Comment (jcrist) [inline distributed/utils_test.py:751]: I can confirm that (at least locally) this change fixes the distributed tests in the `dask/dask` repo.

**Review by jcrist — APPROVED** [review]

**Review by jakirkham — APPROVED** [review]

**Comment by jakirkham** [comment]
Seeing 2 test failures. One with [`test_restart` on Linux Python 3.9]( https://github.com/dask/distributed/runs/5683627158?check_suite_focus=true#step:11:1613 ) and one with [`test_restart_sync_no_center` on Windows Python 3.9]( https://github.com/dask/distributed/runs/5683627382?check_suite_focus=true#step:11:1239 ).

The former seems to be due to a missing file and the latter due to contention around a file.

Not seeing other issues around these tests on the issue tracker (so don't know if they are known to be flaky). Though wouldn't be surprised if these are due to sporadic test failures. Did see PR ( https://github.com/dask/distributed/pull/5994 ) modernizing `test_restart_sync_no_center` (maybe something similar is needed for `test_restart`?).

Idk if these should be blockers, but ok going ahead without them (if others agree) to get Python 3.10 included.

File: distributed/tests/test_profile.py [hunk 9]
```diff
@@ -200,3 +205,110 @@ def stop():
...
+)
+def test_info_frame_f_lineno(f_lasti: int, f_lineno: int) -> None:
+    assert info_frame(
+        FakeFrame(
+            f_lasti=f_lasti,
+            f_code=FakeCode(
+                co_filename="<stdin>",
+                co_name="example",
+                co_firstlineno=1,
+                co_lnotab=b"\x00\x01\x0c\x01\x08\x01\x04\xfe",
```
Comment (fjetter) [inline distributed/tests/test_profile.py:258]: where do these bytes come from? (same question below about `co_code`)

**Comment by mrocklin** [comment]
test_restart_sync_no_center can be nuked.  That's happening at #5994 .  Taking a look at the test_restart report now.

**Comment by mrocklin** [comment]
For test_restart I think it's not related to this PR.  That being said, I think that folks in this PR might be able to help diagnose what's going on.  I think that that failing test (and others) might boil down to this question: https://github.com/dask/distributed/issues/6001

File: continuous_integration/environment-3.10.yaml [hunk 10]
```diff
@@ -0,0 +1,56 @@
+name: dask-distributed
+channels:
+  - conda-forge
+  - defaults
```
Comment (sjperkins) [inline continuous_integration/environment-3.10.yaml:4]: It doesn't seem possible to test pytorch on Python 3.10 until https://github.com/pytorch/pytorch/issues/66424 is addressed. Perhaps a reminder issue for testing pytorch once it has 3.10 support would be suitable?

File: distributed/tests/test_profile.py [hunk 11]
```diff
@@ -200,3 +205,102 @@ def stop():
...
+        (25, 2),
+        (26, 2),
+        (27, 2),
+        (100, 2),
+    ],
+)
+def test_call_stack_f_lineno(f_lasti: int, f_lineno: int) -> None:
+    assert call_stack(FakeFrame(f_lasti=f_lasti, f_code=FAKE_CODE)) == [
+        f'  File "<stdin>", line {f_lineno}, in example\n\t'
+    ]
```
Comment (sjperkins) [inline distributed/tests/test_profile.py:306]: For my own comprehension, custom line number handling for Python frames is necessary because

1. None line numbers have greater frequency in Python 3.10 frames.
2. This causes increased memory usage in distributed logs.
3. Python frame objects are read-only, so it's necessary to mock their internals. I think dataclasses are fine but for e.g. https://naleraphael.github.io/blog/posts/devlog_create_a_builtin_frame_object/  uses ctypes

File: distributed/node.py [hunk 12]
```diff
@@ -131,12 +131,9 @@ def start_http_server(
             import ssl
 
             ssl_options = ssl.create_default_context(
-                cafile=tls_ca_file, purpose=ssl.Purpose.SERVER_AUTH
+                cafile=tls_ca_file, purpose=ssl.Purpose.CLIENT_AUTH
```
Comment (sjperkins) [inline distributed/node.py:134]: From https://docs.python.org/3/library/ssl.html#ssl.Purpose.SERVER_AUTH

> Purpose.SERVER_AUTH
>
> Option for [create_default_context()](https://docs.python.org/3/library/ssl.html#ssl.create_default_context) and [SSLContext.load_default_certs()](https://docs.python.org/3/library/ssl.html#ssl.SSLContext.load_default_certs)[](https://docs.python.org/3/library/ssl.html#ssl.Purpose.CLIENT_AUTH). This value indicates that the context may be used to authenticate web servers (therefore, it will be used to create client-side sockets).
>
> Purpose.CLIENT_AUTH
>
> Option for [create_default_context()](https://docs.python.org/3/library/ssl.html#ssl.create_default_context) and [SSLContext.load_default_certs()](https://docs.python.org/3/library/ssl.html#ssl.SSLContext.load_default_certs). This value indicates that the context may be used to authenticate web clients (therefore, it will be used to create server-side sockets).

I understand this change removes the `ssl.SSLError` but as I don't have a good grasp of SSL and security, I think I should nit here and ask what this change does in greater detail.

**Review by sjperkins — COMMENTED** [review]
Some observations:

* Pytorch is no longer tested on Python 3.10.
* The default SSL Context purpose has changed from `SERVER_AUTH` to `CLIENT_AUTH`
* Do the docs build OK?

File: distributed/tests/test_profile.py (hunk 11 above)
Comment (graingert) [reply distributed/tests/test_profile.py:306]: > This causes increased memory usage in distributed logs.

No the profiler just crashed when the frame was missing an f_lineno

I wanted to use a dataclass so I could control the co_lnotab and co_lines()

File: distributed/node.py (hunk 12 above)
Comment (graingert) [reply distributed/node.py:134]: `HTTPServer` creates server-side sockets and so needs an SSLContext of the `ssl.Purpose.CLIENT_AUTH` kind. so this should always have been `purpose=ssl.Purpose.CLIENT_AUTH`

also CLIENT_AUTH sets
```python
            ssl_options.check_hostname = False
            ssl_options.verify_mode = ssl.CERT_NONE
```
already so we don't need those overrides anymore

File: continuous_integration/environment-3.10.yaml (hunk 10 above)
Comment (fjetter) [reply continuous_integration/environment-3.10.yaml:4]: Good catch. Typically we have three different test builds.

- One cythonized, mostly vanilla dependencies
- One vanilla dependency build
- One with all optional build dependencies

In the past these were distributed on py3.7, 3.8 and 3.9, s.t. python 3.9 installed all dependencies.

While this test matrix still has flaws, it's a pragmatic approach and I'd like to keep it this way.
If the optional dependencies didn't make it to py3.10, yet, we should test them in py3.9 and use the py3.10 as the vanilla build, i.e. the environment.yaml should be similar to the 3.8 one

**Review by sjperkins — APPROVED** [review]

**Comment by mrocklin** [comment]
I don't think that this PR should block on that test

**Comment by jakirkham** [comment]
@fjetter did you have more thoughts here or is this good from your perspective? 🙂

**Comment by fjetter** [comment]
The environment.yaml still includes too many dependencies. All the "only tested here" dependencies should be removed since they are tested in the py3.9 build

File: continuous_integration/environment-3.10.yaml [hunk 13]
```diff
@@ -0,0 +1,56 @@
...
+  - coverage<6.3  # https://github.com/nedbat/coveragepy/issues/1310
+  - dask  # overridden by git tip below
+  - filesystem-spec  # overridden by git tip below
+  - h5py
+  - ipykernel
+  - ipywidgets
+  - jinja2
+  - joblib  # overridden by git tip below
+  - jupyter_client
+  - lz4  # Only tested here
```
Comment (graingert) [inline continuous_integration/environment-3.10.yaml:22]:
```suggestion
```

File: continuous_integration/environment-3.10.yaml [hunk 14]
```diff
@@ -0,0 +1,56 @@
...
+  - msgpack-python
+  - netcdf4
+  - paramiko
+  - pre-commit
+  - prometheus_client
+  - psutil
+  - pynvml  # Only tested here
```
Comment (graingert) [inline continuous_integration/environment-3.10.yaml:29]:
```suggestion
```

File: continuous_integration/environment-3.10.yaml [hunk 15]
```diff
@@ -0,0 +1,56 @@
...
+  - pytest
+  - pytest-cov
+  - pytest-faulthandler
+  - pytest-repeat
+  - pytest-rerunfailures
+  - pytest-timeout
+  - python-blosc  # Only tested here
+  - python-snappy  # Only tested here
```
Comment (graingert) [inline continuous_integration/environment-3.10.yaml:37]:
```suggestion
```

**Review by graingert — COMMENTED** [review]
remove "# Only tested here" deps

File: continuous_integration/environment-3.10.yaml [hunk 16]
```diff
@@ -0,0 +1,52 @@
...
+  - scipy
+  - sortedcollections
+  - tblib
+  - toolz
+  - tornado=6
+  - zict  # overridden by git tip below
+  - zstandard
+  - pip:
+      - git+https://github.com/dask/dask
+      - git+https://github.com/dask/s3fs
```
Comment (fjetter) [inline continuous_integration/environment-3.10.yaml:46]:
```suggestion
```

File: continuous_integration/environment-3.10.yaml [hunk 17]
```diff
@@ -0,0 +1,52 @@
...
+      - git+https://github.com/dask/zict
+      # FIXME https://github.com/dask/distributed/issues/5345
+      # - git+https://github.com/intake/filesystem_spec
+      - git+https://github.com/joblib/joblib
```
Comment (fjetter) [inline continuous_integration/environment-3.10.yaml:50]:
```suggestion
```

File: continuous_integration/environment-3.10.yaml [hunk 18]
```diff
@@ -0,0 +1,52 @@
...
+      - keras
```
Comment (fjetter) [inline continuous_integration/environment-3.10.yaml:51]:
```suggestion
```

File: continuous_integration/environment-3.10.yaml [hunk 19]
```diff
@@ -0,0 +1,52 @@
...
+  - toolz
+  - tornado=6
+  - zict  # overridden by git tip below
+  - zstandard
+  - pip:
+      - git+https://github.com/dask/dask
+      - git+https://github.com/dask/s3fs
+      - git+https://github.com/dask/zict
+      # FIXME https://github.com/dask/distributed/issues/5345
+      # - git+https://github.com/intake/filesystem_spec
```
Comment (fjetter) [inline continuous_integration/environment-3.10.yaml:49]:
```suggestion
```

File: continuous_integration/environment-3.10.yaml [hunk 20]
```diff
@@ -0,0 +1,52 @@
...
+  - click
+  - cloudpickle
+  - coverage<6.3  # https://github.com/nedbat/coveragepy/issues/1310
+  - dask  # overridden by git tip below
+  - filesystem-spec  # overridden by git tip below
+  - h5py
+  - ipykernel
+  - ipywidgets
+  - jinja2
+  - joblib  # overridden by git tip below
```
Comment (fjetter) [inline continuous_integration/environment-3.10.yaml:20]:
```suggestion
```

This is not even part of the repo anymore

**Review by fjetter — COMMENTED** [review]
I'm also wondering why we are testing
- paramiko
- scikit-learn/scipy
- prometheus
- netcdf4
- h5py
- ...

in this build but we can clean up these build files in a follow up PR.

**Review by fjetter — APPROVED** [review]
